src/Microsoft.ML.Data/Utilities/PathUtils.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace Microsoft.ML.Runtime.Data.Utilities
+{
+    internal static class PathUtils
+    {
+        /// <summary>
+        /// Make a full path realtive to a base path.
+        /// </summary>
+        /// <param name="basepath">The base path, assumed to be a directory.</param>
+        /// <param name="path">The full path.</param>
+        /// <returns>The relative path.</returns>
+        /// <exception cref="ArgumentException">If the paths are not relative.</exception>
+        internal static string MakePathRelative(string basepath, string path)
+        {
+            Contracts.AssertNonEmpty(basepath);
+            Contracts.AssertNonEmpty(path);
+
+            Uri baseUri = new Uri(basepath);
+            Uri uri = new Uri(path);
+
+            if (baseUri.Scheme != uri.Scheme)
+            {
+                throw new ArgumentException("Paths cannot be made relative as they are of different schemes.");
+            }
+
+            string relativePath;
+            try
+            {
+                if (!baseUri.AbsoluteUri.EndsWith("/"))
+                {
+                    baseUri = new Uri(baseUri.AbsoluteUri + "/");
+                }
+
+                relativePath = baseUri.MakeRelativeUri(uri).ToString();
+            }
+            catch (ArgumentNullException e)
+            {
+                throw new ArgumentException("Paths could not be made relative.", e);
+            }
+            catch (InvalidOperationException e)
+            {
+                throw new ArgumentException("Paths could not be made relative.", e);
+            }
```

Comment:
We use Contracts.ExceptUserArg for something that is a user argument, with the first parameter being the name of that argumnet. We use Contracts.ExceptParam for something that indicates a fault in a parameter, that is, a method parameter.

---
In reply to: [190345650](https://github.com/dotnet/machinelearning/pull/61#discussion_r190345650) [](ancestors = 190345650)

Reply:
Added Contracts.Except. The difference between the methods are really just the signature and types of exceptions being thrown. Because I have an inner exception, Contracts.Except makes the most sense.